overlord/servicestate/resources/quotas.go

```diff
@@ -0,0 +1,105 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+* Copyright (C) 2021 Canonical Ltd
+*
+* This program is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License version 3 as
+* published by the Free Software Foundation.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*
+ */
+
+package resources
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/gadget/quantity"
+)
+
+// QuotaResourceMemory is the memory limit for the quota group being controlled,
+// either the initial limit the group is created with for the "create"
+// action, or if non-zero for the "update" the memory limit, then the new
+// value to be set.
+type QuotaResourceMemory struct {
+	MemoryLimit quantity.Size
+}
+
+type QuotaResources struct {
+	Memory *QuotaResourceMemory
+}
+
+func (qr *QuotaResources) validateMemoryQuota() error {
+	// make sure the memory limit is not zero
+	if qr.Memory.MemoryLimit == 0 {
+		return fmt.Errorf("cannot create quota group with no memory limit set")
+	}
+
+	// make sure the memory limit is at least 4K, that is the minimum size
+	// to allow nesting, otherwise groups with less than 4K will trigger the
+	// oom killer to be invoked when a new group is added as a sub-group to the
+	// larger group.
+	if qr.Memory.MemoryLimit <= 4*quantity.SizeKiB {
+		return fmt.Errorf("memory limit %d is too small: size must be larger than 4KB", qr.Memory.MemoryLimit)
+	}
+	return nil
+}
+
+func (qr *QuotaResources) Validate() error {
+	if qr.Memory == nil {
+		return fmt.Errorf("quota group must have a memory limit set")
+	}
+
+	if qr.Memory != nil {
+		if err := qr.validateMemoryQuota(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (qr *QuotaResources) ValidateChange(newLimits QuotaResources) error {
+
+	// check that the memory limit is not being decreased
+	if newLimits.Memory != nil && newLimits.Memory.MemoryLimit != 0 {
+		// we disallow decreasing the memory limit because it is difficult to do
+		// so correctly with the current state of our code in
+		// EnsureSnapServices, see comment in ensureSnapServicesForGroup for
+		// full details
+		if qr.Memory != nil && newLimits.Memory.MemoryLimit < qr.Memory.MemoryLimit {
+			return fmt.Errorf("cannot decrease memory limit, remove and re-create it to decrease the limit")
+		}
+	}
+
+	return nil
+}
+
+func (qr *QuotaResources) Change(newLimits QuotaResources) error {
+	if err := qr.ValidateChange(newLimits); err != nil {
+		return err
+	}
+
+	if newLimits.Memory != nil {
+		qr.Memory = newLimits.Memory
```

Comment:
why the `if`? Cannot you change qr.Memory to nil? If that is the case, shouldn't that be validated in `ValidateChange()`? If that method does that, then this could be just `qr = newLimits`.

Reply:
Yes, you are exactly correct, it is because we do not want to set the qr.Memory to nil. The reason the code looks like this is because of the additional logic that is added in this function in the next PR. I can change it for this PR if there is preference for that, and then redo it in the next PR? I'd be glad to do it

Reply:
Tbh I'm not sure if I like the idea of having not nil/nil to know which limit has changed or not, and then only set the changed limits, but I'll wait for the other PR to see exactly how it has been implemented :-)

Reply:
One alternative in this situation is to keep a bool that specifies whether a limit has changed or not, and I'm not sure I personally would prefer that. It makes more sense from a API point of view to not change a limit that hasn't been supplied (think json api requests), and to represent that we either need an accompanioning boolean value that indicates whether the limit is set, or we can represent it using a pointer type.

so that alternative would look like:

newLimits.IsMemoryValid
newLimits.Memory

but personally I don't feel like that is more elegant.